packages/firestore/src/local/instance_metadata_channel.ts

```diff
@@ -0,0 +1,455 @@
+      );
+      if (instanceState) {
+        this.activeInstances[instanceState.instanceKey] = instanceState;
+      }
+    }
+
+    this.started = true;
+    this.persistState();
+  }
+
+  getMinimumGloballyPendingMutation(): BatchId | null {
+    let minMutationBatch = null;
+    Object.keys(this.activeInstances).forEach(key => {
+      if (minMutationBatch === null) {
+        minMutationBatch = this.activeInstances[key].minMutationBatchId;
+      } else {
+        minMutationBatch = Math.min(
+          this.activeInstances[key].minMutationBatchId,
+          minMutationBatch
+        );
+      }
+    });
+
+    return minMutationBatch;
+  }
+
+  getGloballyActiveQueryTargets(): SortedSet<TargetId> {
+    let activeTargets = new SortedSet<TargetId>(primitiveComparator);
+
+    Object.keys(this.activeInstances).forEach(key => {
+      activeTargets = activeTargets.unionWith(
+        this.activeInstances[key].activeTargetIds
+      );
+    });
+
+    return activeTargets;
+  }
+
+  addLocalPendingMutation(batchId: BatchId): void {
+    this.localInstanceState.addPendingMutation(batchId);
+    this.persistState();
+  }
+
+  removeLocalPendingMutation(batchId: BatchId): void {
+    this.localInstanceState.removePendingMutation(batchId);
+    this.persistState();
+  }
+
+  addLocallyActiveQueryTarget(targetId: TargetId): void {
+    this.localInstanceState.addQueryTarget(targetId);
+    this.persistState();
+  }
+
+  removeLocallyActiveQueryTarget(targetId: TargetId): void {
+    this.localInstanceState.removeQueryTarget(targetId);
+    this.persistState();
+  }
+
+  shutdown(): void {
+    assert(
+      this.started,
+      'LocalStorageMetadataNotifier.shutdown() called when not started'
+    );
+    window.removeEventListener('storage', this.storageListener);
+    this.localStorage.removeItem(this.storageKey);
+    this.started = false;
+  }
+
+  private handleStorageEvent(event: StorageEvent): void {
+    if (!this.started) {
+      return;
+    }
+    if (event.storageArea === this.localStorage) {
+      // TODO(multitab): This assert will likely become invalid as we add garbage
+      // collection.
+      assert(
+        event.key !== this.storageKey,
+        'Received LocalStorage notification for local change.'
+      );
+
+      const instanceKey = fromLocalStorageKey(event.key, this.instanceKeyRe, 3);
+      if (instanceKey) {
+        if (event.newValue == null) {
+          delete this.activeInstances[instanceKey];
+        } else {
+          const newInstance = InstanceState.fromLocalStorageEntry(
+            instanceKey,
+            event.newValue
+          );
+          if (newInstance) {
+            this.activeInstances[newInstance.instanceKey] = newInstance;
+          }
+        }
+      }
+    }
+  }
+
+  private get localInstanceState(): LocalInstanceState {
+    return this.activeInstances[this.instanceKey] as LocalInstanceState;
+  }
+
+  private persistState(): void {
```

Comment:
I think we should consider aggregating state changes to avoid too much thrashing if the user does a bunch of operations back-to-back (writing to local storage tons of times and other tabs parsing the new entries tons of times). So rather than write to localStorage immediately, we could e.g. delay it by 100ms and batch up all the persistState() calls within that window into 1.  This would actually alleviate a lot of my performance concerns.

Consider adding a TODO(multitab)?

Reply:
Added todo